File: skills/code-architect/references/code-examples/go/mediator.go
// Mediator: centralise the communication between components so they refer
// to the mediator instead of to each other directly.
package main

import "fmt"

// Mediator coordinates colleagues; they never talk to each other directly.
type Mediator interface {
	Notify(sender string, event string)
}

// colleague holds a back-reference to its mediator.
type colleague struct {
	name      string
	mediator  Mediator
	lastHeard string
}

func (c *colleague) Send(event string) {
	c.mediator.Notify(c.name, event)
}

// hub is the concrete mediator wiring colleagues together.
type hub struct {
	members map[string]*colleague
}

func (h *hub) register(c *colleague) {
	c.mediator = h
	h.members[c.name] = c
}

func (h *hub) Notify(sender, event string) {
	for name, c := range h.members {
		if name != sender {
			c.lastHeard = sender + ":" + event
		}
	}
}

func main() {
	h := &hub{members: map[string]*colleague{}}
	a := &colleague{name: "A"}
	b := &colleague{name: "B"}
	h.register(a)
	h.register(b)

	a.Send("ping")
	fmt.Println("B heard:", b.lastHeard)
}

File: skills/code-architect/references/code-examples/rust/mediator.rs
// Mediator — centralize communication between components so they refer only to
// the mediator instead of to each other, reducing coupling.

use std::cell::RefCell;
use std::rc::Rc;

// The mediator coordinates colleagues by name.
struct Hub {
    log: RefCell<Vec<String>>,
}

impl Hub {
    fn new() -> Self {
        Hub { log: RefCell::new(Vec::new()) }
    }

    // A colleague notifies the hub, which decides what others should do.
    fn broadcast(&self, sender: &str, event: &str) {
        self.log.borrow_mut().push(format!("{} -> {}", sender, event));
        match (sender, event) {
            ("Switch", "on") => println!("Hub: turning Lamp ON, starting Fan"),
            ("Switch", "off") => println!("Hub: turning Lamp OFF, stopping Fan"),
            _ => println!("Hub: no rule for {}:{}", sender, event),
        }
    }
}

struct Switch {
    hub: Rc<Hub>,
}

impl Switch {
    fn toggle(&self, on: bool) {
        self.hub.broadcast("Switch", if on { "on" } else { "off" });
    }
}

fn main() {
    let hub = Rc::new(Hub::new());
    let switch = Switch { hub: Rc::clone(&hub) };

    switch.toggle(true);
    switch.toggle(false);

    println!("event log: {:?}", hub.log.borrow());
}
